dlt/extract/incremental/__init__.py

```diff
@@ -560,8 +561,14 @@ def __call__(self, rows: TDataItems, meta: Any = None) -> Optional[TDataItems]:
         else:
             rows = self._transform_item(transformer, rows)
 
-        # write back state
-        self._cached_state["last_value"] = transformer.last_value
+        # Skip writing back state if no valid rows for last_value in transformer may be wrong
+        if rows is not None:
```

Comment:
this is really a good catch but the problem is deeper than this. we cannot let the state go back and indeed when there's a lag this can unfortunately happen (like in your test, where the "last value" record is missing in the second read).

so IMO this code will fix it
```py
         if self.lag and "last_value" in self._cached_state:
            transformer.last_value = self.last_value_func(transformer.last_value, self._cached_state["last_value"])
        self._cached_state["last_value"] = transformer.last_value
```
it uses last_value_func to make sure that "highest" last_value is preserved ("highest" per order established by the last_value_func)

`rows` is of TDataItems type: so it may be a single item or a list of items. items may be dicts, panda frames, arrow tables or batches.

so for example rows may be a pandas frame or a list of pandas frames. that's way the checks below are not sufficient. but as I said the problem is deeper (but solution is at the end simpler)

Reply:
Thanks, @rudolfix! I think this modification aligns well with the dlt code style. As you mentioned, there might be deeper issues that I’m not aware of at the moment.

>so for example rows may be a pandas frame or a list of pandas frames. that's way the checks below are not sufficient. but as I said the problem is deeper (but solution is at the end simpler)

I have a question. If rows is an empty list of pandas DataFrames, I think the code `rows != []` will handle this condition. This is because of the following code:

```
        if isinstance(rows, list):
            rows = [
                item
                for item in (self._transform_item(transformer, row) for row in rows)
                if item is not None
            ]
        else:
            rows = self._transform_item(transformer, rows)
```